File: Assets/scripts/ship.js
//#pragma strict
// Stores the state of ship systems, and the progress with hyperspace calculations.
// Ship can continue calculations as long as the computer is working, but if the critical systems,
// (for now, the engine and computer) aren't working, then the ship won't jump until they are fixed)
// jumping requires pressing the "jump to lightspeed button" (does not automatically jump)


// for now, keeps track of engine and computer explicitly as the critical systems.
var evasionChance : int = 0;
var calculationTimeRemaining : float = 60.0;
var engine : GameObject; 
var computer : GameObject; 
var currentMission : GameObject;
var components;
var damageableComponents;
var criticalComponents; // both are arrays of components. critical components must be intact to jump.
var rooms; // all rooms in ship 

function Start() {
	components = new Array();
	damageableComponents = new Array();
	criticalComponents = new Array();
	rooms = new Array();
	currentMission = GameObject.FindGameObjectWithTag("Mission");
}


function Update () {
	//TODO Refactor this!
	if (computer.GetComponent("resourceConsumer").working) {
		calculationTimeRemaining -= Time.deltaTime;
		if (calculationTimeRemaining < 0) {
			calculationTimeRemaining = 0;
			currentMission.SendMessage("ReadyToJump");
		}
	}
}

function HyperspaceJump () {
	if (calculationTimeRemaining <= 0) {
		var engineWorking = engine.GetComponent("resourceConsumer").working;
		var computerWorking = computer.GetComponent("resourceConsumer").working;
		if (engineWorking && computerWorking) {
			Debug.Log("Jumped to Hyperspace!");
			// complete mission here.
			currentMission.SendMessage("CompleteMission");
		}
	} 
}

function IncomingDamage() {
	// performs calculations to see if ship dodges event (called from mission object)
	Debug.Log("Damage Called, attempting to dodge");
	var hitRoll = Random.Range(0.0, 100.0);
	if (hitRoll > evasionChance) {
		// then the ship is hit, call take Damage
		TakeDamage();
	} else {
		Debug.Log("Damage Avoided!");
	}
}

// pick room, then check children, and randomly select conduit for damage
function TakeDamage() {
	var damagedUnit : int = Random.Range(0, damageableComponents.length - 1);
	var damage : int = Random.Range(10, 100);
	damageableComponents[damagedUnit].SendMessage("TakeDamage", damage);
	Debug.Log("component damaged: " + damageableComponents[damagedUnit] + ", for " + damage + " damage.");
//	for (var component : GameObject in components) {
//		
//	}
}

function AddShipComponent(component : GameObject) {
	components.Push(component);
	// add special components to slots.
	var type :ComponentType = component.GetComponent("shipComponent").componentType;
	var damageable : damageableComponent = component.GetComponent("damageableComponent");
	if (type == ComponentType.engine) {
		engine = component;
	} else if (type == ComponentType.computer) {
		computer = component;
	}
	
	if (damageable != null) {
		damageableComponents.Push(component);
		Debug.Log("added damageableComponent: " + component);
	}
	
//	Debug.Log("adding component: " + component);
}

function AddShipCriticalComponent(component : GameObject) {
	AddShipComponent(component);
	criticalComponents.Push(component);
}

function AddEvasion(ammount : int) {
	evasionChance += ammount;
	evasionChance = CheckBounds(evasionChance, 0, 100);
}

function RemoveEvasion(ammount : int) {
	evasionChance -= ammount;
	evasionChance = CheckBounds(evasionChance, 0, 100);
}

function CheckBounds(param : int, min : int, max : int) {
	if (param < min) {
		param = min;
	} else if (param > max) {
		param = max;
	}
	
	return param;
}

File: Assets/scripts/c src/Ship.cs
using UnityEngine;
using System.Collections;
using System.Collections.Generic;

public class Ship : MonoBehaviour {

	public float evasionChance = 0.0f;
	public float calculationTimeRemaining = 60.0f;
	public GameObject computer;
	private ResourceComponent computerScript;
	public GameObject currentMission;
	public List<GameObject> components = new List<GameObject>();
	public List<GameObject> damageableComponents = new List<GameObject>();
	public List<GameObject> criticalComponents = new List<GameObject>(); // both are arrays of components. critical components must be intact to jump.

	// for displaying hit notifications
	private bool displayHitNotification = false;
	private string hitMessage = "";

	// Use this for initialization
	void Start () {
		currentMission = GameObject.FindGameObjectWithTag("Mission");
	}
	
	// Update is called once per frame
	void Update () {
		//TODO Refactor this!
		if (calculationTimeRemaining > 0 && computerScript.isActive) {
			calculationTimeRemaining -= Time.deltaTime * computerScript.GetComponentOutput();
			if (calculationTimeRemaining < 0) {
				calculationTimeRemaining = 0;
				currentMission.SendMessage("ReadyToJump");
			}
		}
	}

	void HyperspaceJump () {
		if (calculationTimeRemaining <= 0) {
			bool allCriticalSystemsWorking = true;
			foreach (GameObject component in criticalComponents) {
				if (!component.GetComponent<ResourceComponent>().isActive) {
					allCriticalSystemsWorking = false;
					break;
				}
			}
			if (allCriticalSystemsWorking) {
				Debug.Log("Jumped to Hyperspace!");
				// complete mission here.
				currentMission.SendMessage("CompleteMission");
			}
		} 
	}

	void IncomingDamage() {
		// performs calculations to see if ship dodges event (called from mission object)
		Debug.Log("Damage Called, attempting to dodge");
		float hitRoll = Random.Range(0.0f, 100.0f);
		if (hitRoll > evasionChance) {
			// then the ship is hit, call take Damage
			TakeDamage();
		} else {
			Debug.Log("Damage Avoided!");
		}
	}
	
	// pick room, then check children, and randomly select conduit for damage
	void TakeDamage() {
		int damagedUnit = Random.Range(0, damageableComponents.Count);
		float damage = Random.Range(10.0f, 100.0f);
		damageableComponents[damagedUnit].SendMessage("TakeDamage", damage);

		SetHitMessage (damageableComponents[damagedUnit], damage);
		StartCoroutine ("DisplayHitMessage");

		Debug.Log("component damaged: " + damageableComponents[damagedUnit] + ", for " + damage + " damage.");
		//	for (var component : GameObject in components) {
		//		//TODO damage multiple components at once
		//	}
	}

	void AddShipComponent(GameObject component) {
		components.Add(component);
		// add special components to slots.
		ShipComponent shipComponent = component.GetComponent<ShipComponent>();
		DamageableComponent damageable = component.GetComponent<DamageableComponent>();

		if (shipComponent.isComputer) {
			computer = component;
			computerScript = component.GetComponent<ResourceComponent>();
		} 
		
		if (damageable != null) {
			damageableComponents.Add(component); // TODO should the damageable component itself be added here instead?
			Debug.Log("added damageableComponent: " + component);
		}
		
		//	Debug.Log("adding component: " + component);
	}

	void AddShipCriticalComponent(GameObject component) {
		AddShipComponent(component);
		criticalComponents.Add(component);
	}


	void AddEvasion(float amount) {
		evasionChance += amount;
		evasionChance = CheckBounds(evasionChance, 0.0f, 100.0f);
		Debug.Log ("Adding evasion: " + amount);
	}
	
	void RemoveEvasion(float amount) {
		evasionChance -= amount;
		evasionChance = CheckBounds(evasionChance, 0.0f, 100.0f);
	}
	
	float CheckBounds(float param, float min, float max) {
		if (param < min) {
			param = min;
		} else if (param > max) {
			param = max;
		}
		
		return param;
	}

	void SetHitMessage(GameObject hit, float amount) {
		hitMessage = hit.name + " was hit for " + amount + " damage!";
	}

	IEnumerator DisplayHitMessage() {
		displayHitNotification = true;
			yield return new WaitForSeconds(2.0f);
		displayHitNotification = false;

	}

	// ============================
	void OnGUI () {
		GUI.skin.box.wordWrap = true;
		if (displayHitNotification) {
			GUI.Box (new Rect (0, Screen.height - 50, 300, 100), hitMessage);
		}
	}


}
